common/cuda_hip/matrix/csr_kernels.hpp.inc

```diff
@@ -842,4 +842,125 @@ __global__
+            values[iz + row_start] *= beta[0];
+            if (col_idxs[iz + row_start] == row) {
+                values[iz + row_start] += alpha[0];
+            }
+        }
+    }
+}
+
+
+template <typename IndexType>
+__global__ __launch_bounds__(default_block_size) void check_diagonal_entries(
+    const IndexType num_min_rows_cols,
+    const IndexType* const __restrict__ row_ptrs,
+    const IndexType* const __restrict__ col_idxs,
+    int* const __restrict__ block_reduced)
+{
+    constexpr int warp_size = config::warp_size;
+    auto warp_tile =
+        group::tiled_partition<warp_size>(group::this_thread_block());
+    const auto warpid = thread::get_subwarp_id_flat<warp_size, IndexType>();
+    const auto num_warps = thread::get_subwarp_num_flat<warp_size, IndexType>();
+    constexpr int warps_per_block = default_block_size / warp_size;
+    __shared__ int has_diag[warps_per_block];
+    for (int i = threadIdx.x; i < warps_per_block; i += blockDim.x) {
+        has_diag[i] = 1;
+    }
+    __syncthreads();
+    const int warpid_in_block = threadIdx.x / warp_size;
+    for (IndexType row = warpid; row < num_min_rows_cols; row += num_warps) {
+        const auto tid_in_warp = warp_tile.thread_rank();
+        const IndexType row_start = row_ptrs[row];
+        const IndexType num_nz = row_ptrs[row + 1] - row_start;
+        int row_has_diag_local = 0;
+        for (IndexType iz = tid_in_warp; iz < num_nz; iz += warp_size) {
+            if (col_idxs[iz + row_start] == row) {
+                row_has_diag_local = 1;
+                break;
+            }
+        }
+        const config::lane_mask_type row_has_diag =
+            warp_tile.ballot(row_has_diag_local);
+        if (warp_tile.thread_rank() == 0) {
+            has_diag[warpid_in_block] =
+                static_cast<int>(has_diag[warpid_in_block] && row_has_diag);
+        }
+    }
+    __syncthreads();
+    if (warpid_in_block == 0) {
+        // accumulate into one warp length
+        for (uint32 i = 1; i <= warps_per_block / warp_size; i <<= 1) {
+            if (i * warp_size + warp_tile.thread_rank() < warps_per_block) {
+                has_diag[warp_tile.thread_rank()] = static_cast<int>(
+                    has_diag[warp_tile.thread_rank()] &&
+                    has_diag[i * warp_size + warp_tile.thread_rank()]);
+            }
+        }
+        // warp-reduce
+        int val = warp_tile.thread_rank() < warps_per_block
+                      ? has_diag[warp_tile.thread_rank()]
+                      : 1;
+        // TODO: round up warps_per_block to a power of 2 below
+        // (by default it already is)
+        for (uint32 k = static_cast<uint32>(warps_per_block / 2); k >= 1;
+             k >>= 1) {
+            val = static_cast<int>(val && warp_tile.shfl_down(val, k));
+        }
+        if (warp_tile.thread_rank() == 0) {
+            block_reduced[blockIdx.x] = val;
+        }
+    }
```

Comment:
I don't see an `is_sorted_by_row`. Which one are you referring to? The algorithm for `is_sorted_by_col` does not have coalesced access; this might actually be better. Eventually, I think we will move to provide the user with an option of disabling this check, as Pratik mentioned somewhere above. That could be done either through user-settable matrix properties (have a property called `has_all_diagonal_entries`), or using the checking infrastructure that Gregor had started working on.

Reply:
ah yeah, that's the one I mean. is_sorted_by_col_idxs provides fast-fail semantics due to the fact that you can short-circuit the && operator. Also as the result is either `true` or `false`, and if it is `false` in one thread, that already determines the global result, you can simplify that a lot like `if (warp.any(!has_diagonal)) { if (lane == 0) { flag = false; } return; }`